turbinia/task_manager.py

```diff
@@ -181,42 +192,164 @@ def add_evidence(self, evidence_):
+      bool: Indicating whether we are done.
     """
     return not bool(len(self.tasks))
 
+  def check_request_done(self, request_id):
+    """Checks if we have any outstanding tasks for the request ID.
+
+    Args:
+      request_id (str): The request ID to check for completion
+
+    Returns:
+      bool: Indicating whether all Jobs are done.
+    """
+    job_completion = []
+    for job in self.running_jobs:
+      if request_id == job.request_id:
+        job_completion.append(job.check_done())
+
+    return min(job_completion)
+
+  def check_request_finalized(self, request_id):
+    """Checks if the the request is done and finalized.
+
+    A request can be done but not finalized if all of the Tasks created by the
+    original Jobs have completed, but the "finalize" Job/Tasks have not been
+    run.  These finalize Job/Tasks are created after all of the original
+    Jobs/Tasks have completed. Only one Job needs to be marked as finalized for
+    the entire request to be considered finalized.
+
+    Args:
+      request_id (str): The request ID to check for finalization.
+
+    Returns:
+      bool: Indicating whether all Jobs are done.
+    """
+    request_finalized = False
+    for job in self.running_jobs:
+      if request_id == job.request_id and job.is_finalized:
+        request_finalized = True
+        break
+
+    return request_finalized and self.check_request_done(request_id)
+
   def get_evidence(self):
     """Checks for new evidence to process.
 
     Returns:
-      A list of Evidence objects
+      list[evidence.Evidence]: The evidence to process.
     """
     raise NotImplementedError
 
-  def add_task(self, task, evidence_):
+  def get_job(self, job_id):
+    """Gets the running Job instance from the given Job ID
+
+    Args:
+      job_id (str): The Job id to get the job for.
+
+    Returns:
+      TurbiniaJob|None: Job instance if found, else None
+    """
+    job = None
+    for job_instance in self.running_jobs:
+      if job_id == job_instance.id:
+        job = job_instance
+        break
+
+    return job
+
+  def generate_request_finalize_tasks(self, job):
+    """Generates the Tasks to finalize the given request ID.
+
+    Args:
+      job (TurbiniaJob): The last Job that was run for this request.
+    """
+    request_id = job.request_id
+    final_job = jobs_manager.JobsManager.GetJobInstance('FinalizeRequestJob')
+    final_job.request_id = request_id
+    final_job.evidence.config = job.evidence.config
+    log.debug(
+        'Request {0:s} done, but not finalized, creating FinalizeRequestJob '
+        '{1:s}'.format(request_id, final_job.id))
+
+    # Finalize tasks use EvidenceCollection with all evidence created by the
+    # request or job.
+    final_evidence = evidence.EvidenceCollection()
+    final_evidence.request_id = request_id
+    self.running_jobs.append(final_job)
+
+    # Gather evidence created by every Job in the request.
+    for running_job in self.running_jobs:
```

Comment:
So the jobs are still in self.running_jobs even when all their tasks are done?

Reply:
Ah, I see, so the running_jobs are still populated when the finalize task is added. Jobas are removed after the Finalize task har run. Correct?

Reply:
Yeah, that's correct :)